src/Data/Vec/Relation/Binary/Lex/Core.agda

```diff
@@ -0,0 +1,122 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Lexicographic ordering of vectors
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Vec.Relation.Binary.Lex.Core where
+
+open import Data.Empty
+open import Data.Nat using (ℕ)
+open import Data.Product using (_×_; _,_; proj₁; proj₂; uncurry)
+open import Data.Sum using (_⊎_; inj₁; inj₂; [_,_])
+open import Data.Vec using (Vec; []; _∷_)
+open import Data.Vec.Relation.Binary.Pointwise.Inductive using (Pointwise; []; _∷_)
+open import Function.Base using (flip)
+open import Function.Equivalence using (_⇔_; equivalence)
+open import Relation.Binary
+open import Relation.Binary.PropositionalEquality as P using (_≡_)
+open import Relation.Nullary hiding (Irrelevant)
+import Relation.Nullary.Decidable as Dec
+open import Relation.Nullary.Product using (_×-dec_)
+open import Relation.Nullary.Sum using (_⊎-dec_)
+open import Relation.Nullary.Negation
+open import Level using (Level; _⊔_)
+
+-- The lexicographic ordering itself can be either strict or non-strict,
+-- depending on whether type P is inhabited.
+
+data Lex {a ℓ₁ ℓ₂} {A : Set a} (P : Set) (_≈_ : Rel A ℓ₁) (_≺_ : Rel A ℓ₂) : ∀ {m n} → REL (Vec A m) (Vec A n) (a ⊔ ℓ₁ ⊔ ℓ₂) where
+  base : (p : P) → Lex P _≈_ _≺_ [] []
+  this : ∀ {x y m n} {xs : Vec A m} {ys : Vec A n} → (x≺y : x ≺ y) → (m≡n : m ≡ n) → Lex P _≈_ _≺_ (x ∷ xs) (y ∷ ys)
+  next : ∀ {x y m n} {xs : Vec A m} {ys : Vec A n} → (x≈y : x ≈ y) → (xs<ys : Lex P _≈_ _≺_ xs ys) → Lex P _≈_ _≺_ (x ∷ xs) (y ∷ ys)
+
+Lex-mapP : ∀ {a ℓ₁ ℓ₂} {A : Set a} {P₁ P₂ : Set} {_≈_ : Rel A ℓ₁} {_≺_ : Rel A ℓ₂} → ∀ {m n} {xs : Vec A m} {ys : Vec A n} → (P₁ → P₂) → Lex P₁ _≈_ _≺_ xs ys → Lex P₂ _≈_ _≺_ xs ys
+Lex-mapP f (base p) = base (f p)
+Lex-mapP f (this x≺y m≡n) = this x≺y m≡n
+Lex-mapP f (next x≈y xs<ys) = next x≈y (Lex-mapP f xs<ys)
+
+-- Properties
+
+module _ {a ℓ₁ ℓ₂} {A : Set a} {P : Set} {_≈_ : Rel A ℓ₁} {_≺_ : Rel A ℓ₂} where
+
+  private
+    _≋_ = Pointwise _≈_
+    _<_ = Lex P _≈_ _≺_
+
+  length-equal : ∀ {m n P} {xs : Vec A m} {ys : Vec A n} → Lex P _≈_ _≺_ xs ys → m ≡ n
```

Comment:
If this is provable, why do we have the length polymorphism in the first place?

Reply:
For the same reason that permutations have 2 length parameters too: the proof isn't always `refl` and it's not always definitionally `refl`.  Things compose a whole lot better if you don't conflate two things that are equal because it's a non-trivial property. [At least, that's my guess.]

Reply:
I think it was mostly because pointwise did it and I assumed that design choices there were reasonable (but I assume @JacquesCarette is correct).

Reply:
Another question is if it is desirable: I was expecting e.g. [0, 1, 2] < [0, 1, 3, 4, 5].

Reply:
That's tricky. It's kind of lifting the domain to always have a (virtual) infimum, even when it doesn't. I'm not saying it's wrong, just requires thought. And, of course, the above doesn't do that. I think you'd at least need to drop the `m≡n` term from `this` to get that behaviour.

Reply:
I agree with @fredrikNordvallForsberg that [0,1,2] < [0, 1, 3, 4, 5] should be true. As does our implementation of [lexicographic orderings for lists](https://github.com/agda/agda-stdlib/blob/0f437abd5d4c4e9b35429ce6a3f66ff2c708f373/src/Data/List/Relation/Binary/Lex/Core.agda#L34).

Reply:
Nice!

Reply:
We should maybe have both?

* lexicographic for Vec-as-lists
* lexicographic for Vec-as-(homogeneous)-n-tuples

Now to find names...

Reply:
My thinking was that if you order Vec-as-lists you can use the list in the ffirst place (or at least bijection to lists).

Reply:
> My thinking was that if you order Vec-as-lists you can use the list in the ffirst place (or at least bijection to lists).

That's a fair point :+1: 

> Now to find names...

From reading the [wikipedia page](https://en.wikipedia.org/wiki/Lexicographical_order) on the various versions, I'd suggest `Lex.Cartesian.(Strict/NonStrict)` for this variant where lengths have to be the same, `Lex.Dictionary.(Strict/NonStrict)` for the existing variant under `List` where the shorter sequences can be below in the order and `Lex.Short.(Strict/NonStrict)` for the third variant which apparently exists where shorter sequences are always below in the order.

Thoughts?

Anyway now that I think that we're sure we're keeping this one, I'll go through and add some minor comments on the naming.